src/check_executor.rs

```diff
@@ -0,0 +1,243 @@
+        status,
+        scheduled_check_time,
+        actual_check_time,
+        duration,
+        status_reason,
+        ..
+    } = result;
+
+    let status_label = match status {
+        CheckStatus::Success => "success",
+        CheckStatus::Failure => "failure",
+        CheckStatus::MissedWindow => "missed_window",
+    };
+    let failure_reason = match status_reason.as_ref().map(|r| r.status_type) {
+        Some(CheckStatusReasonType::Failure) => Some("failure"),
+        Some(CheckStatusReasonType::DnsError) => Some("dns_error"),
+        Some(CheckStatusReasonType::Timeout) => Some("timeout"),
+        None => None,
+    };
+
+    // Record duration of check
+    if let Some(duration) = duration {
+        metrics::histogram!(
+            "check_result.duration_ms",
+            "histogram" => "timer",
+            "status" => status_label,
+            "failure_reason" => failure_reason.unwrap_or("ok"),
+        )
+        .record(duration.num_milliseconds() as f64);
+    }
+
+    // Record time between scheduled and actual check
+    metrics::histogram!(
+        "check_result.delay_ms",
+        "histogram" => "timer",
+        "status" => status_label,
+        "failure_reason" => failure_reason.unwrap_or("ok"),
+    )
+    .record((*actual_check_time - *scheduled_check_time).num_milliseconds() as f64);
+
+    // Record status of the check
+    metrics::counter!(
+        "check_result.processed",
+        "status" => status_label,
+        "failure_reason" => failure_reason.unwrap_or("ok"),
+    )
+    .increment(1);
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{task::Poll, time::Duration};
+
+    use chrono::Utc;
+    use futures::poll;
+    use similar_asserts::assert_eq;
+    use tokio::{sync::oneshot::Receiver, time};
+    use uuid::Uuid;
+
+    use super::*;
+    use crate::{
+        checker::dummy_checker::DummyChecker, producer::dummy_producer::DummyResultsProducer,
+    };
+
+    #[tokio::test(start_paused = true)]
+    async fn test_executor_simple() {
+        let checker = Arc::new(DummyChecker::new(Duration::from_secs(1)));
+        let producer = Arc::new(DummyResultsProducer::new("uptime-results"));
+
+        let (sender, _) = run_executor(1, checker, producer);
+
+        let tick = Tick::from_time(Utc::now());
+        let config = Arc::new(CheckConfig {
+            subscription_id: Uuid::from_u128(1),
+            ..Default::default()
+        });
+
+        let resolve_rx = queue_check(&sender, tick, config.clone());
+        tokio::pin!(resolve_rx);
+
+        // Will not be resolved yet
+        time::sleep(Duration::from_millis(100)).await;
+        assert_eq!(poll!(resolve_rx.as_mut()), Poll::Pending);
+
+        let result = resolve_rx.await.unwrap();
+        assert_eq!(result.subscription_id, config.subscription_id);
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn test_executor_concurrent_limit() {
+        let checker = Arc::new(DummyChecker::new(Duration::from_secs(1)));
```

Comment:
Should we just make this 100ms so that test isn't too slow?

Reply:
Time is frozen, so it auto advances

Reply:
Lol right, I keep forgetting this